core-api/src/telemetry/metrics.rs

```diff
@@ -0,0 +1,250 @@
+#[derive(Clone, Debug)]
+pub struct LangMetricAttributes {
+    /// A set of references to attributes stored in lang memory. All referenced attributes should
+    /// be attached to the metric when recording.
+    pub ids: HashSet<u64>,
+    /// If populated, these key values should also be used in addition to the referred-to
+    /// existing attributes when recording
+    pub new_attributes: Vec<MetricKeyValue>,
+}
+
+impl MetricAttributes {
+    /// Extend existing metrics attributes with others, returning a new instance
+    pub fn merge(&self, other: MetricAttributes) -> Self {
+        let mut me = self.clone();
+        match (&mut me, other) {
+            (MetricAttributes::OTel { ref mut kvs }, MetricAttributes::OTel { kvs: other_kvs }) => {
+                Arc::make_mut(kvs).extend((*other_kvs).clone());
+            }
+            (MetricAttributes::Lang(ref mut l), MetricAttributes::Lang(ol)) => {
+                l.ids.extend(ol.ids);
+                l.new_attributes.extend(ol.new_attributes);
+            }
+            _ => panic!("Cannot merge metric attributes of different kinds"),
+        }
+        me
+    }
+
+    /// Mutate self to add new kvs
+    pub fn add_new_attrs(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        match self {
+            #[cfg(feature = "otel_impls")]
+            MetricAttributes::OTel { ref mut kvs, .. } => {
+                Arc::make_mut(kvs).extend(new_kvs.into_iter().map(Into::into));
+            }
+            MetricAttributes::Lang(ref mut attrs, ..) => {
+                attrs.new_attributes.extend(new_kvs.into_iter());
+            }
+        }
+    }
+}
+
+/// Options that are attached to metrics on a per-call basis
+#[derive(Clone, Debug, Default, derive_more::Constructor)]
+pub struct MetricsAttributesOptions {
+    pub attributes: Vec<MetricKeyValue>,
+}
+impl MetricsAttributesOptions {
+    pub fn extend(&mut self, new_kvs: impl IntoIterator<Item = MetricKeyValue>) {
+        self.attributes.extend(new_kvs.into_iter())
+    }
+}
+
+/// A K/V pair that can be used to label a specific recording of a metric
+#[derive(Clone, Debug)]
+pub struct MetricKeyValue {
+    pub key: String,
+    pub value: MetricValue,
+}
+impl MetricKeyValue {
+    pub fn new(key: impl Into<String>, value: impl Into<MetricValue>) -> Self {
+        Self {
+            key: key.into(),
+            value: value.into(),
+        }
+    }
+}
+
+/// Values metric labels may assume
+#[derive(Clone, Debug, derive_more::From)]
+pub enum MetricValue {
+    String(String),
+    Int(i64),
+    Float(f64),
+    Bool(bool),
+    // can add array if needed
+}
+impl From<&'static str> for MetricValue {
+    fn from(value: &'static str) -> Self {
+        MetricValue::String(value.to_string())
+    }
+}
+
+pub trait Counter: Send + Sync {
+    fn add(&self, value: u64, attributes: &MetricAttributes);
+}
+
+pub trait Histogram: Send + Sync {
+    // When referring to durations, this value is in millis
```

Comment:
Is there a case where we use histograms for anything but durations? In the Go SDK, we were able to get away with three metric types: counter, gauge, and timer (which is histogram, but we called it timer and clearly used durations). I wonder if a similar thing would make sense here.

I am worried that people will want to know whether you are recording a duration or a numeric value.

Reply:
They very easily could be non-durations, and either way it's a histogram. The language is clear with "when" - I don't think this is an issue. At the end of the day it's still a histogram regardless of what unit you've chosen.

Reply:
But I think it's helpful for our users to know whether they're recording a duration or not. We can say "may be milliseconds or just some other number" but they they're gonna have some extra conditionals checking string name to know it's a duration. I think we should have a separate concept for duration-based histograms and non-duration-based histograms (and if there are only duration-based histograms ever used, we can wait until we need the non-duration-based to add it).

Reply:
I disagree. There are basically zero other metrics APIs that do this. There are metric types, which generally correlate to how things are aggregated, and there are units. Confusing units with the method of aggregation is I think not an appropriate design.

If we would like to accept units, sure, that's totally reasonable, but IMO it's a definite no-no to consider two metric types to be different simply because they are using different units.

Reply:
> If we would like to accept units, sure, that's totally reasonable, but IMO it's a definite no-no to consider two metric types to be different simply because they are using different units.

:+1: I will happily take units. I just figured separate traits was the best way to represent separate units instead of int + enum or something. I was never different metric types on the downstream system (they both still become the same histograms). For example, with Prometheus in some languages people create either a "timer" or a "histogram", but the former is just sugar over the latter with the type as duration.

Reply:
Cool. I will update things to allow passing units and descriptions in when building instruments. They are just wrappers around strings at the end of the day.

Reply:
:+1: If at all possible would like an enumerate or similar for units instead of an arbitrary value. For any lang-side impl of this, they can abstract them to different overloads (I suspect only .NET will be fast enough to support a lang-side impl).

Reply:
Certainly possible but the problem is it only has the meaning we give it. It ends up becoming a string, at least using OTel as a backend. I can make one up, but then users are limited only to what we've made up, where they might have some unit they care about we can never imagine... So I think we should just keep it a string.

Reply:
> users are limited only to what we've made up

This is what drove the metrics abstraction choices in Go. We didn't want a general purpose metrics abstraction for all uses, we wanted a Temporal metrics abstraction for only how Temporal uses metrics. Yes, this limits users reusing that abstraction inside of their workflow/activity to just what we also use, but what we've needed has been enough so far and if/when we offer custom metrics impls, they will use their more powerful impl instead of the weaker Temporal abstraction in super advanced cases.

I prefer the strongly typed int or duration. If we can't have that, I prefer the strongly typed int with strongly typed unit enum. If we can't have that, I guess I can settle for int with arbitrary string, but it's the least preferred.

Reply:
They're still limited to the kinds we expose, which seems to me the important part. Units are just a string description exported to the backend. Limiting users to what that string can be doesn't seem to help them in any material way. If the language layer wants to offer some "TimingHistogram" abstraction, it is still absolutely free to do that, and adding a typed enum on the core side can't meaningfully enforce anything interesting, we just don't gain anything. I can't somehow enforce that when you pick a `Milliseconds` variant, that the number you're passing actually makes any sense.

Reply:
> Units are just a string description exported to the backend

That assumes a lot about the backend. We always prefer strongly typed, even if the backend is stringly typed. For example, an implementation may want to record durations and let a config decide the duration unit. Or they may have to translate (for example some backends use milliseconds and [some use seconds](https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewTimer) by default).

> If the language layer wants to offer some "TimingHistogram" abstraction, it is still absolutely free to do that, and adding a typed enum on the core side can't meaningfully enforce anything interesting, we just don't gain anything.

:+1: Can we at least document all the string units _we_ set for this value so that users can convert your value back to a duration to have the freedom of deciding unit?

> adding a typed enum on the core side can't meaningfully enforce anything interesting

It just signals intent. That you can only pass a duration to the `TimingHistogram` makes it clearer what the value really is. Kinda the same reason these languages even have duration types instead of just integer types even though they technically can be expressed as integers in their greatest-supported resolution.

Reply:
FYIW I feels like Prometheus encourage us to take the wrong direction, and we should not follow it's lead simply on the principle that "most engines out there" are doing it that way. Metric emitters should focus on what those values really means (ie to them), rather than how they will be stored or presented.

The fact is that some place in the code is either emitting time values, or something else, but in the end, the unit for that metric is unambiguous for the caller and should therefore also be clearly specified as part the "metric" contract. The data is a duration, expressed in ms. The fact that it will likely be presented as an histogram is of lesser significancy.

Reply:
> FYIW I feels like Prometheus encourage us to take the wrong direction, and we should not follow it's lead simply on the principle that "most engines out there" are doing it that way.

It's not about a specific backend, that was just an example. It's not taking me in a certain direction. Very simply, we should type our datatypes as specifically as is useful. Knowing that something is a duration and not just an integer is useful. People need to translate durations.

>  the unit for that metric is unambiguous for the caller

Not true if someone has to `== "milliseconds"` in their code. That's pretty ambiguous. Hope we don't add "seconds" next.

>  The data is a duration, expressed in ms.

More like the data is an integer, expressed with an arbitrary string you must know to assume ms.

Reply:
> Can we at least document all the string units we set
Core doesn't even set any units, right now, but I can add that to this PR.

Anyway, my main point here is it doesn't really make much sense to expose the unit type safety stuff at the Core level, because lang can do that, and the translation to calls into core is quite easy to get right. I agree of course that if users want to specify unit & associated value types in lang that that makes sense.

The only thing I could do in Core would maybe be something like accepting `std::time::Duration` for some kinda time-specific histogram, but even then I'm just immediately going to toss that out the window when I make the real metric call, and lang should've already done that it's type.

Reply:
> The only thing I could do in Core would maybe be something like accepting std::time::Duration for some kinda time-specific histogram, but even then I'm just immediately going to toss that out the window when I make the real metric call,

But I, the language dev, am not going to toss that out the window. I am going to give it to the user's metric impl and they're gonna be happy I differentiated.

But we've discussed enough, string is good enough I suppose. Let's just document the string core uses for its durations so I can have a hardcoded conditional for stronger typing lang side.